api/lib/controllers/meeting.js

```diff
@@ -280,14 +297,70 @@ module.exports = {
         },
         { $unwind: '$meetings' },
         { $replaceRoot: { newRoot: '$meetings' } },
+        ...pipelineFilters,
         { $sort: { 'date': -1 } },
         { $skip: parseInt( skip ) },
         { $limit: parseInt( limit ) }
       ];
 
+      const totalCount = [ //this is more performant than $facet
```

Comment:
If we add the `pipeline_filters` to the `pipeline` of each `$lookup` of the `meetings` collection we can then use `$size` to count the length of `meetings` before the `$skip` and `$limit` steps. We could then either sort the array using array opts then take a slice of it to simulate skip and limit or we could just add a property to each of the meetings returned which is `$size` and perform a final `$map` to return `{ meetings: [], total: 30 }`. This would likely be more performant than making a second query to get `totalCount`.

Reply:
@thudsonbu I just realized when we get to a lot larger datasets, I think that this solution of saving the entire list to memory before sorting limiting and skipping the array will be less performant than a second query. Im not sure what you mean by 

>add a property to each of the meetings returned which is $size and perform a final $map"

Reply:
@zbarnz You are right that loading all the meetings into memory and performing the aggregation could be expensive. However, the amount of meetings that a user would need to have in order for it to be more expensive than a network request would need to be in the hundreds of thousands. An intra zone network request (between our mongodb cluster and agenda) takes between one and five milliseconds while copying 5000 meetings roughly 64kb of data into main memory takes around 5 milliseconds . The CPU operations take less than a microsecond but the data would need to be copied around a few times. Even if we go on the high end for those manipulations at 100 microseconds it should still be between 50 times faster compared to the one to five millisecond network request.

We could make the two requests concurrently so that they both traverse the network at the same time but the first operation would block the second (and potentially request from other users) on the mongo side and it still loads all the meetings into memory since it is an aggregation. So thats why I am hoping for single query.

As for the poorly worded sentence, what I meant is that we could add the `count` to each meeting document in an additional `$project`. On second thought, it would probably be better to perform the `$skip` and `$limit` on the `meetings` array that exists before the `$unwind` and add the `count` to the single document before the `$unwind` as well leaving us with something like this:

```js
const sliceStart = parseInt(skip);
const sliceEnd = parseInt(limit) + sliceStart - 1;

const pipeline = [
  {
    $match: { _id: ObjectID(subject_id) },
  },
  {
    $lookup: {
      from: "meetings",
      localField: "_id",
      foreignField: "owner_id",
      pipeline: pipelineFilters,
      as: "owned_meetings",
    },
  },
  {
    $lookup: {
      from: "participants",
      pipeline: [
        { $match: { email: subject_email } },
        {
          $lookup: {
            from: "meetings",
            localField: "meeting_id",
            foreignField: "_id",
            pipeline: pipelineFilters,
            as: "meetings",
          },
        },
        {
          $project: {
            meeting: { $arrayElemAt: ["$meetings", 0] },
          },
        },
        {
          $replaceRoot: {
            newRoot: "$meeting",
          },
        },
      ],
      as: "participating_meetings",
    },
  },
  {
    $project: {
      meetings: {
        $concatArrays: ["$participating_meetings", "$owned_meetings"],
      },
    },
  },
  {
    $project: {
      meetings: {
        $sortArray: {
          input: "$meetings",
          sortBy: { date: -1 },
        },
      },
      total: { $size: "$meetings" },
    },
  },
  {
    $project: {
      meetings: {
        $slice: ["$meetings", sliceStart, sliceEnd],
      },
      total: 1,
    },
  },
];

const [{ meetings, total }] = await User.aggregate(pipeline);
```